### DISCUSSION TITLE
Changing ESP Algorithms after rekeying

### DISCUSSION THREAD

**Discussion by EAFedorov** [strongswan/strongswan#2471] [General]
Hi.
I use strongswan 5.9.14. I did some testing and got a strange result.

I noticed that the ESP algorithms change after rekeing.

I have a central server and several remote servers that connect to it.
I don't set the list of ESP-algorithms on remote servers. I want to receive it from the central server.

1. On the server I use the following settings:

```
lifetime = 5m
esp = aes256-sha1
```

If I use encryption and integrity algorithms in the server settings, then these algorithms are selected.

```
Sep 26 11:51:36 server charon: 13[CFG] received proposals: ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Sep 26 11:51:36 server charon: 13[CFG] configured proposals: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ, ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Sep 26 11:51:36 server charon: 13[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
```

During rekeing, the remote server sends only used algorithms list, and they are selected again.

```
Sep 26 11:54:40 server charon: 07[CFG] received proposals: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
Sep 26 11:54:40 server charon: 07[CFG] configured proposals: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ, ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Sep 26 11:54:40 server charon: 07[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ 
```

2. If I use the AEAD algorithm in the server settings, then this algorithm is selected.

lifetime = 5m
esp = aes256gcm16

```
Sep 26 11:55:50 server charon: 13[CFG] received proposals: ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Sep 26 11:55:50 server charon: 13[CFG] configured proposals: ESP:AES_GCM_16_256/NO_EXT_SEQ, ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Sep 26 11:55:50 server charon: 13[CFG] selected proposal: ESP:AES_GCM_16_256/NO_EXT_SEQ
```

But, During the rekeing, the remote server sends only one set of encryption and integrity algorithms and does not send the one in use.
As a result, the set of algorithms changes.

```
Sep 26 11:58:51 server charon: 01[CFG] received proposals: ESP:AES_CBC_128/HMAC_SHA2_256_128/NO_EXT_SEQ
Sep 26 11:58:51 server charon: 01[CFG] configured proposals: ESP:AES_GCM_16_256/NO_EXT_SEQ, ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Sep 26 11:58:51 server charon: 01[CFG] selected proposal: ESP:AES_CBC_128/HMAC_SHA2_256_128/NO_EXT_SEQ
```

The result is as follows:
If I set the encryption and integrity algorithms in ipsec.conf, they are selected after each rekeing.
If I set the AEAD algorithm, it will change after the first rekeing to the default encryption and integrity algorithms of the remote server.

Is this correct behavior?

**Comment by tobiasbrunner** [comment]
> Is this correct behavior?

No, not sending the proposal that's currently in use doesn't make much sense. What IKEv2 implementation is running on the remote servers?

**Comment by EAFedorov** [comment]
Hello.
All servers use strongswan 5.9.14.
The ike settings do not affect the situation in any way.
I checked. I used different settings in ike, but the AEAD algoritm in esp always change after the first rekeing, if they are not set on the remote server.

**Reply by tobiasbrunner** [reply]
Just to clarify, you are using the same **unmodified** 5.9.14 release (e.g. built from the release tarball or the release tag) on all hosts? Because I can't reproduce the behavior with that version. I always see all the proposals in the rekey request.

**Comment by EAFedorov** [comment]
Hello.
Yes. I am using the same unmodified 5.9.14 release from this link 
https://download.strongswan.org/strongswan-5.9.14.tar.gz
I unpacked and run the commands: ./configere, make, sudo make install

I did some more tests.

I see algorithms changing only when using ipsec.conf and the key lifetime on the central server ends earlier than on the remote server.

Here are my settings
Central server:
```
config setup
	charondebug = "ike 1, lib 1, cfg 2, knl 2"
conn %default
	keyexchange = ikev2
	ikelifetime = 24h
	lifetime = 5m
	rekeymargin = 1m
	ike = aes256-sha1-modp2048
	esp = aes256gcm16
	left = x.x.x.x
	leftauth = psk
	auto = add
conn test
	right = x.x.x.y
	rightauth = psk
```
Remote server:
```
config setup
	charondebug = "ike 1, lib 1, cfg 2, knl 2"
conn %default
	keyexchange = ikev2
	ikelifetime = 24h
	lifetime = 1h
	rekeymargin = 5m
	left = x.x.x.y
	leftauth = psk
	auto = add
conn test
	right = x.x.x.x
	rightauth = psk
```
The remote server initiates a connection and the aes256gcm16 agrees in the ESP.
Central server log:
```
Oct  2 14:29:41 server charon: 04[CFG]   candidate "test" with prio 5+5
Oct  2 14:29:41 server charon: 04[CFG] found matching child config "test" with prio 10
Oct  2 14:29:41 server charon: 04[CFG] selecting proposal:
Oct  2 14:29:41 server charon: 04[CFG]   no acceptable ENCRYPTION_ALGORITHM found
Oct  2 14:29:41 server charon: 04[CFG] selecting proposal:
Oct  2 14:29:41 server charon: 04[CFG]   proposal matches
Oct  2 14:29:41 server charon: 04[CFG] received proposals: ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Oct  2 14:29:41 server charon: 04[CFG] configured proposals: ESP:AES_GCM_16_256/NO_EXT_SEQ, ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Oct  2 14:29:41 server charon: 04[CFG] selected proposal: ESP:AES_GCM_16_256/NO_EXT_SEQ 
```
Remote server log:
```
Oct  2 14:29:41 server charon: 15[CFG] selecting proposal:
Oct  2 14:29:41 server charon: 15[CFG]   no acceptable ENCRYPTION_ALGORITHM found
Oct  2 14:29:41 server charon: 15[CFG] selecting proposal:
Oct  2 14:29:41 server charon: 15[CFG]   proposal matches
Oct  2 14:29:41 server charon: 15[CFG] received proposals: ESP:AES_GCM_16_256/NO_EXT_SEQ
Oct  2 14:29:41 server charon: 15[CFG] configured proposals: ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Oct  2 14:29:41 server charon: 15[CFG] selected proposal: ESP:AES_GCM_16_256/NO_EXT_SEQ
```
But, after 5 minutes (lifetime on the central server) the AEAD algorithm changes to encryption and integrity algorithms
Central server log:
```
Oct  2 14:33:01 server charon: 02[CFG] selecting proposal:
Oct  2 14:33:01 server charon: 02[CFG]   no acceptable ENCRYPTION_ALGORITHM found
Oct  2 14:33:01 server charon: 02[CFG] selecting proposal:
Oct  2 14:33:01 server charon: 02[CFG]   proposal matches
Oct  2 14:33:01 server charon: 02[CFG] received proposals: ESP:AES_CBC_128/HMAC_SHA2_256_128/NO_EXT_SEQ
Oct  2 14:33:01 server charon: 02[CFG] configured proposals: ESP:AES_GCM_16_256/NO_EXT_SEQ, ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Oct  2 14:33:01 server charon: 02[CFG] selected proposal: ESP:AES_CBC_128/HMAC_SHA2_256_128/NO_EXT_SEQ
```
Remote server log:
```
Oct  2 14:33:01 server charon: 03[CFG] selecting proposal:
Oct  2 14:33:01 server charon: 03[CFG]   no acceptable ENCRYPTION_ALGORITHM found
Oct  2 14:33:01 server charon: 03[CFG] selecting proposal:
Oct  2 14:33:01 server charon: 03[CFG]   proposal matches
Oct  2 14:33:01 server charon: 03[CFG] received proposals: ESP:AES_GCM_16_256/NO_EXT_SEQ, ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Oct  2 14:33:01 server charon: 03[CFG] configured proposals: ESP:AES_CBC_128/AES_CBC_192/AES_CBC_256/HMAC_SHA2_256_128/HMAC_SHA2_384_192/HMAC_SHA2_512_256/HMAC_SHA1_96/AES_XCBC_96/NO_EXT_SEQ, ESP:AES_GCM_16_128/AES_GCM_16_192/AES_GCM_16_256/NO_EXT_SEQ
Oct  2 14:33:01 server charon: 03[CFG] selected proposal: ESP:AES_CBC_128/HMAC_SHA2_256_128/NO_EXT_SEQ 
```
**The algorithm changes only when using AEAD in ipsec.conf and the lifetime on the central server is less than on the remote server.**

If I use any encryption and integrity algorithms, they do not change after rekeing.
If the lifetime on the remote server is less than on the central server, then the algorithms do not change.
If I use swanctl, then the algorithms do not change.

**Reply by tobiasbrunner** [reply]
Ah, I see. The `received proposals: ESP:AES_CBC_128/HMAC_SHA2_256_128/NO_EXT_SEQ` message threw me off, which is actually logged by the initiator of the rekeying after receiving the response. From your original description, I thought it was logged by the  responder, which I couldn't reproduce.

The algorithm switch you observe is normal behavior because strongSwan prefers its own proposal over that of the remote peer, by default. This can be changed by disabling `charon.prefer_configured_proposals` in [strongswan.conf](https://docs.strongswan.org/docs/5.9/config/strongswanConf.html). Or you can also change the central server configuration so it doesn't append the default proposal by configuring `esp = aes256gcm16!`.

**Comment by EAFedorov** [comment]
Thank you.
